mlbench_core/optim/pytorch/optim.py

```diff
@@ -462,6 +480,109 @@ def step(self, closure=None):
             closure (callable, optional): A closure that reevaluates the model
                 and returns the loss.
         """
-        self.agg(self.model, self.agg_mode)
         loss = super(CentralizedAdam, self).step(closure=closure)
+        self.agg(self.model, self.agg_mode)
         return loss
+
+
+class FP32Optimizer:
+    """
+    Standard optimizer, computes backward and applies weight update.
+    """
+
+    def __init__(
+        self, model, optimizer, grad_clip=None, world_size=1, average_models=True
+    ):
+        """
+        Constructor for the Fp32Optimizer
+
+        :param grad_clip: coefficient for gradient clipping, max L2 norm of the
+            gradients
+        """
+        self.model = model
+        self.params = model.parameters()
+        self.grad_clip = grad_clip
+        self.optimizer = optimizer
+        self.agg = AllReduceAggregation(world_size=world_size).agg_grad
+        if average_models:
+            self.agg_mode = "avg"
+        else:
+            raise NotImplementedError("Only average model is supported right now.")
+
+    def step(self):
+        """
+        Performs one step of the optimizer.
+
+        :param loss: value of loss function
+        :param optimizer: optimizer
+        :param update: if True executes weight update
+        """
+        self.agg(self.model, self.agg_mode)
+        if self.grad_clip != float("inf"):
+            clip_grad_norm_(self.params, self.grad_clip)
+        return self.optimizer.step()
+
+    def backward_loss(self, loss):
+        loss.backward()
+
+    def zero_grad(self):
+        self.optimizer.zero_grad()
+
+
+class AMPOptimizer:
+    """
+    Optimizer compatible with AMP.
+    Uses AMP to apply loss scaling, computes backward and applies weight
+    update.
+    """
+
+    def __init__(
+        self,
+        model,
+        optimizer,
+        grad_clip=None,
+        loss_scale=8192,
+        dls_upscale_interval=128,
+        world_size=1,
+        average_models=True,
+    ):
+        """
+        Constructor for the AMPOptimizer
+
+        :param model: model
+        :param grad_clip: coefficient for gradient clipping, max L2 norm of the
+            gradients
+        """
+        self.model = model
+        self.grad_clip = grad_clip
+        self.optimizer = optimizer
+        loss_scaler = amp._amp_state.loss_scalers[0]
+        loss_scaler._loss_scale = loss_scale
+        loss_scaler._scale_seq_len = dls_upscale_interval
+        self.agg = AllReduceAggregation(
+            world_size=world_size, cast_in=torch.float32, cast_out=torch.float16
```

Comment:
BTW, I found [horovod](https://github.com/horovod/horovod) can support FP16 gradient reduction via their modified MPI. maybe it is worth to have a [look](https://github.com/horovod/horovod/blob/4371007e154b7882062cab8696b26d520252971b/test/test_torch.py#L179).

Reply:
Okay, I will check this.